### PR TITLE
[serve] Don't use mixin class for class-based backends

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Currently, we require users to use the `RayServeMixin` in order to create a backend from a class. However, we need control over the backend `__init__` in order to implement fault tolerance and there is no guarantee that a mixin's `__init__` will be called, so we need to change the approach. This instead wraps the user-defined class, so they can just pass the class definition in instead of using the mixin (also a bit of an API improvement).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by edoakes** [comment]
@simon-mo do I need to update any documentation?

**Comment by simon-mo** [comment]
No doc changes needed for now (of course, make sure to add docstring)

File: python/ray/serve/backend_worker.py [hunk 1]
```diff
@@ -12,23 +12,53 @@
...
+            return self.backend.get_metrics()
 
-        # This parameter let argument inspection work with inner function.
-        self.__wrapped__ = func_to_run
+        async def handle_request(self, request):
+            return await self.backend.handle_request(request)
 
-    def __call__(self, *args, **kwargs):
-        return self.func(*args, **kwargs)
+    RayServeWrappedWorker.__name__ = "RayServeWorker_" + func_or_class.__name__
```
Comment (simon-mo) [inline python/ray/serve/backend_worker.py:60]: nice!

File: python/ray/serve/backend_worker.py [hunk 2]
```diff
@@ -48,108 +78,82 @@ def ensure_async(func):
...
-        self._serve_metric_latency_list = []
+class RayServeWorker:
+    """Fetches requests and handles them with the provided callable."""
+
+    def __init__(self,
+                 name,
+                 _callable,
+                 self_handle,
+                 router_handle,
+                 is_function=False):
```
Comment (simon-mo) [inline python/ray/serve/backend_worker.py:89]: this should be positional? it should always be passed in explicitly right? it doesn't make sense to me to assume a value here

File: python/ray/serve/backend_worker.py [hunk 3]
```diff
@@ -243,28 +245,6 @@ def _ray_serve_count_num_positional(self, f):
         # re-assign to default values
         serve_context.web = False
         serve_context.batch_size = None
-
-        # Tell router that current actor is idle
-        self._ray_serve_fetch()
+        self.mark_idle_in_router()
```
Comment (simon-mo) [inline python/ray/serve/backend_worker.py:248]: we should remove this and change to router --push--> worker to reduce one rpc. the same logic can be applied when the response is successful returned from worker==worker is idle

**Review by simon-mo — APPROVED** [review]
LGTM, make sure to add docstring in necessary places

File: python/ray/serve/backend_worker.py (hunk 2 above)
Comment (edoakes) [reply python/ray/serve/backend_worker.py:89]: Fair point, changed

File: python/ray/serve/backend_worker.py (hunk 3 above)
Comment (edoakes) [reply python/ray/serve/backend_worker.py:248]: Agreed. I will do this in a follow-up

**Comment by edoakes** [comment]
Punting docstrings to next PR because tests are passing here and the next one is dependent.

**Pull request merged**